SalesforceCommerceCloud/pwa-kit
# Create useAppOrigin hook to consume x-forwarded-host header value

# Description
Issues: https://github.com/SalesforceCommerceCloud/pwa-kit/issues/1257

TLDR: Retail React App sometimes uses getAppOrigin to build links, on server side, it returns the value of process.env.EXTERNAL_HOSTNAME which is an MRT hostname and by pass the original host.

## Solution
When CDN is stacked in front of Pwa kit/MRT, the CDN will forward a header `x-forwarded-host` that contains the origin host. To consume this value, we pass it into `res.locals` in the server, and created a React Hook to consume it in the pwa-kit-react-sdk

To make it in a non-breaking fashion, a flag `useXForwardedHost` is introduced in the app configuration (default to false). When it is enabled, getAppOrigin will return the value of x-forwarded-host from request header. 



# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

1. With useXForwardedHost enabled: 
You can't test this locally since it needs to have CDN that forwards the req headers. I've uploaded a bundle to this MRT env `https://www.phased-launch-testing.com/` that has CDN set up as intended
   - Visit https://www.phased-launch-testing.com/. Make sure it is deployed with bundle below
![image](https://github.com/user-attachments/assets/62809ac7-c403-4568-9c53-2803c4c50648)
   - Check the client and server both returns the value of origin host which is `https://www.phased-launch-testing.com` not `https://scaffold-pwa-bjnl-prd.mobify-storefront.com/`
        Client: https://www.phased-launch-testing.com/
        Server https://www.phased-launch-testing.com/?__server_only=1
        
![image](https://github.com/user-attachments/assets/3cf58d85-719f-4b95-9652-0c9b25c86285)

2. With useXForwardedHost disabled: 
Deployed to `https://www.phased-launch-testing.com/` with bundle 
![image](https://github.com/user-attachments/assets/92af7a10-591d-4813-bee2-f49c82c70723)
- After it finished deploying, Visit https://www.phased-launch-testing.com/ and https://www.phased-launch-testing.com?__server_only=1. You will notice getAppOrigin on server will return `https://scaffold-pwa-bjnl-prd.mobify-storefront.com/`



# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
